File: latest/week3/workshop7_m3/W07_M03_demo.py
"""W07-M03: Advanced Annotation in Code — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
os.makedirs("output", exist_ok=True)

# ── 1. LOAD DATA ────────────────────────────────────────
nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year
yearly = nf.query("2015 <= year_added <= 2021").groupby(["year_added","type"]).size().reset_index(name="n")
movies = yearly.query("type == 'Movie'")
tvshows = yearly.query("type == 'TV Show'")

# ── 2. adjustText: NON-OVERLAPPING LABELS ────────────────
np.random.seed(42)
x = np.random.uniform(1, 10, 15); y = np.random.uniform(1, 10, 15)
labels = [f"Point {i+1}" for i in range(15)]

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 5))
# BAD
ax1.scatter(x, y, s=40, c="#1565C0", edgecolors="white", lw=0.5, zorder=5)
for xi, yi, lab in zip(x, y, labels):
    ax1.text(xi+0.1, yi+0.1, lab, fontsize=6)
ax1.set_title("BAD: Overlapping labels\n(ax.text without adjustment)",
    fontweight="bold", color="#C62828")
# GOOD
ax2.scatter(x, y, s=40, c="#1565C0", edgecolors="white", lw=0.5, zorder=5)
try:
    from adjustText import adjust_text
    texts = [ax2.text(xi, yi, lab, fontsize=6) for xi, yi, lab in zip(x, y, labels)]
    adjust_text(texts, ax=ax2,
        arrowprops=dict(arrowstyle="-", color="#888", lw=0.5),
        force_text=0.5, force_points=0.3)
    ax2.set_title("GOOD: adjustText — auto-repelled\n(pip install adjustText)",
        fontweight="bold", color="#1565C0")
except ImportError:
    for i, (xi, yi, lab) in enumerate(zip(x, y, labels)):
        off = 0.4 * (1 if i%2==0 else -1)
        ax2.annotate(lab, (xi,yi), xytext=(xi+off,yi+off), fontsize=6,
            arrowprops=dict(arrowstyle="-", color="#888", lw=0.3))
    ax2.set_title("FALLBACK: manual offset\n(adjustText not installed)",
        fontweight="bold", color="#E65100")
for ax in [ax1,ax2]: ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("Label Overlap: Problem and Solution", fontweight="bold")
plt.tight_layout(); plt.savefig("output/adjusttext_demo.png", dpi=300); plt.close()

# ── 3. CALLOUT BOX WITH CURVED ARROW ────────────────────
peak = movies.loc[movies.n.idxmax()]
latest = movies.iloc[-1]
decline = round((1 - latest["n"] / peak["n"]) * 100)

fig, ax = plt.subplots(figsize=(10, 6))
# Grey context
ax.plot(tvshows.year_added, tvshows.n, "-o", color="#DDDDDD", lw=0.8, markersize=3)
# Story line
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
# Reference line at peak
ax.axhline(y=peak["n"], ls=":", color="#888", lw=0.5)
# Shaded decline region
ax.axvspan(2019.5, 2021.5, alpha=0.03, color="#E53935")
# Callout box with curved arrow
ax.annotate(f"–{decline}% decline\nfrom 2019 peak",
    xy=(2020.5, (peak["n"] + latest["n"]) / 2),
    xytext=(2017.5, latest["n"] * 0.75),
    fontsize=10, fontweight="bold", color="#E53935",
    bbox=dict(boxstyle="round,pad=0.4", facecolor="white",
        edgecolor="#E53935", alpha=0.9, lw=1.5),
    arrowprops=dict(arrowstyle="->",
        connectionstyle="arc3,rad=0.3",
        color="#E53935", lw=1.5))
# Direct labels (no legend)
ax.text(2021.15, movies.iloc[-1]["n"], " Movies", fontsize=9,
    fontweight="bold", color="#1565C0", va="center")
ax.text(2021.15, tvshows.iloc[-1]["n"], " TV Shows", fontsize=8,
    color="#BBBBBB", va="center")
ax.set_xlim(2014.5, 2022.5)
ax.set_title(f"Movie additions declined {decline}% from their 2019 peak",
    fontsize=13, fontweight="bold")
ax.text(0.01, -0.08, "Source: Netflix dataset | UNYT",
    transform=ax.transAxes, fontsize=7, color="#888")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/fully_dressed.png", dpi=300); plt.close()

# ── 4. ANNOTATION LAYERS (progressive build) ────────────
fig, axes = plt.subplots(2, 2, figsize=(12, 8))

for i, (ax, title) in enumerate(zip(axes.flatten(), [
    "Layer 0: Bare chart", "Layer 1: + Reference line",
    "Layer 2: + Direct label", "Layer 3: + Callout box"])):
    ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=1.5, markersize=4)
    ax.set_title(title, fontsize=9, fontweight="bold")
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
    ax.tick_params(labelsize=6)
    if i >= 1:  # reference line
        ax.axhline(y=peak["n"], ls=":", color="#888", lw=0.5)
    if i >= 2:  # direct label
        ax.text(2021.1, latest["n"], " Movies", fontsize=7,
            fontweight="bold", color="#1565C0", va="center")
    if i >= 3:  # callout box
        ax.annotate(f"–{decline}%", xy=(2020, latest["n"]),
            xytext=(2018, latest["n"]*0.8),
            fontsize=9, fontweight="bold", color="#E53935",
            bbox=dict(boxstyle="round,pad=0.2", facecolor="white",
                edgecolor="#E53935"),
            arrowprops=dict(arrowstyle="->", color="#E53935", lw=1))
fig.suptitle("Annotation Hierarchy: Build in Layers", fontweight="bold")
plt.tight_layout(); plt.savefig("output/annotation_layers.png", dpi=300); plt.close()

# ── 5. e-Car: FULLY ANNOTATED CRISIS CHART ──────────────
ec = pd.read_csv("ecar.csv"); ec.columns=[c.strip().replace("  "," ") for c in ec.columns]
ec["Year"] = pd.to_datetime(ec["Approve Date"], format="%m/%d/%Y", errors="coerce").dt.year
ec["Spread"] = ec["Rate"] - ec["Cost of Funds"]
yr = ec.query("2002<=Year<=2012").groupby("Year").agg(rate=("Rate","mean"),spread=("Spread","mean")).reset_index()

fig, ax = plt.subplots(figsize=(10, 5))
ax.axvspan(2007.5, 2009.5, alpha=0.04, color="#E53935")  # crisis window
ax.plot(yr.Year, yr.spread, color="#DDDDDD", lw=0.8)  # grey context
ax.plot(yr.Year, yr.rate, "-o", color="#1565C0", lw=2.5, markersize=5)
# Lehman callout
ax.annotate("Lehman collapse\n(Sep 2008)",
    xy=(2008, yr.query("Year==2008").rate.values[0]),
    xytext=(2005, yr.rate.max()*0.85),
    fontsize=9, fontweight="bold", color="#E53935",
    bbox=dict(boxstyle="round,pad=0.3", facecolor="white", edgecolor="#E53935", lw=1.2),
    arrowprops=dict(arrowstyle="->", connectionstyle="arc3,rad=-0.2",
        color="#E53935", lw=1.5))
# Direct labels
ax.text(2012.2, yr.iloc[-1].rate, " Rate", fontsize=9, fontweight="bold",
    color="#1565C0", va="center")
ax.text(2012.2, yr.iloc[-1].spread, " Spread", fontsize=8, color="#BBB", va="center")
ax.set_xlim(2001.5, 2013.5)
ax.set_title("Loan rates dropped 2pp after the 2008 financial crisis",
    fontsize=12, fontweight="bold")
ax.text(0.01, -0.08, "Source: e-Car dataset | UNYT",
    transform=ax.transAxes, fontsize=7, color="#888")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/ecar_annotated.png", dpi=300); plt.close()

# ── 6. ARROW STYLES DEMO ────────────────────────────────
fig, ax = plt.subplots(figsize=(9, 5))
styles = [("->", "Standard", 0.85), ("-|>", "Filled", 0.70),
    ("-[", "Bar end", 0.55), ("fancy", "Fancy", 0.40), ("wedge", "Wedge", 0.25)]
for style, label, y in styles:
    try:
        ax.annotate(f"  {label} (arrowstyle='{style}')",
            xy=(0.7, y), xytext=(0.2, y), fontsize=9, va="center",
            arrowprops=dict(arrowstyle=style, color="#1565C0", lw=1.5,
                connectionstyle="arc3,rad=0.1"))
    except:
        ax.annotate(f"  {label} ('{style}') — not supported",
            xy=(0.7, y), xytext=(0.2, y), fontsize=9, va="center",
            arrowprops=dict(arrowstyle="->", color="#888", lw=1))
ax.set_xlim(0, 1); ax.set_ylim(0.1, 0.95)
ax.set_title("Arrow Styles: ax.annotate(arrowprops=dict(arrowstyle=...))",
    fontweight="bold")
ax.axis("off")
plt.tight_layout(); plt.savefig("output/arrow_styles.png", dpi=300); plt.close()

print("\nAll W07-M03 Python outputs saved")

File: latest/week3/workshop7_m3/W07_M03_demo.R
# ============================================================
# Workshop 7 — Module 3: Advanced Annotation in Code
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# ── 1. LOAD DATA ────────────────────────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)),
    year_added = year(date_added),
    primary_country = str_trim(str_extract(country, "^[^,]+")))

# ── 2. ggrepel: NON-OVERLAPPING SCATTER LABELS ──────────
library(ggrepel)

top_countries <- nf |>
  filter(!is.na(primary_country), !is.na(year_added),
    year_added >= 2015, year_added <= 2021) |>
  count(primary_country, sort = TRUE) |>
  slice_max(n, n = 15)

p_repel <- ggplot(top_countries,
  aes(x = fct_reorder(primary_country, n), y = n, label = primary_country)) +
  geom_point(color = "#1565C0", size = 3) +
  geom_text_repel(size = 3, max.overlaps = 20,
    segment.color = "#888", segment.size = 0.3,
    box.padding = 0.4, force = 2, seed = 42,
    color = "#333", fontface = "bold") +
  coord_flip() +
  theme_minimal() +
  labs(title = "ggrepel: Auto-positioned, non-overlapping labels",
       subtitle = "15 countries — no manual positioning needed",
       x = NULL, y = "Total Titles (2015–2021)")
ggsave("output/ggrepel_scatter.png", p_repel, width = 9, height = 6, dpi = 300)

# ── 3. ggrepel ON TIME SERIES: Label endpoints ──────────
nf_genres <- nf |> separate_rows(listed_in, sep = ", ") |>
  mutate(genre = str_trim(listed_in))
top5 <- nf_genres |> count(genre, sort = TRUE) |> slice_max(n, n = 5) |> pull(genre)

genre_yearly <- nf_genres |>
  filter(genre %in% top5, !is.na(year_added), year_added >= 2016, year_added <= 2021) |>
  count(year_added, genre)

endpoints <- genre_yearly |>
  group_by(genre) |> filter(year_added == max(year_added))

p_repel_ts <- ggplot(genre_yearly, aes(x = year_added, y = n, color = genre)) +
  geom_line(linewidth = 1) +
  geom_point(data = endpoints, size = 2.5) +
  geom_text_repel(data = endpoints,
    aes(label = genre), direction = "y",
    nudge_x = 0.3, segment.size = 0.3,
    size = 3, fontface = "bold",
    hjust = 0, seed = 42) +
  scale_color_brewer(palette = "Set2") +
  coord_cartesian(xlim = c(2016, 2023)) +
  theme_minimal() +
  theme(legend.position = "none") +
  labs(title = "ggrepel on time series: direct labels at endpoints",
       subtitle = "Replaces legend — labels placed automatically without overlap",
       x = "Year", y = "Titles Added")
ggsave("output/ggrepel_timeseries.png", p_repel_ts, width = 10, height = 5, dpi = 300)

# ── 4. ggtext: MARKDOWN IN TITLES ────────────────────────
# Note: requires ggtext installed
# library(ggtext)
# For demonstration, we simulate the effect with standard theme
yearly_type <- nf |>
  filter(!is.na(year_added), year_added >= 2015, year_added <= 2021) |>
  count(year_added, type)

# Standard version (no ggtext)
p_title <- ggplot(yearly_type, aes(x = year_added, y = n, color = type)) +
  geom_line(linewidth = 1.5) + geom_point(size = 3) +
  scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal() + theme(legend.position = "none") +
  labs(title = "Movies (blue) peaked in 2019; TV Shows (red) grew steadily",
       subtitle = "With ggtext: title words coloured to match — no legend needed")

# ggtext version (uncomment if installed)
# p_title_md <- ggplot(yearly_type, aes(x = year_added, y = n, color = type)) +
#   geom_line(linewidth = 1.5) + geom_point(size = 3) +
#   scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
#   labs(title = "**<span style='color:#1565C0'>Movies</span>** peaked in 2019;
#     **<span style='color:#E53935'>TV Shows</span>** grew steadily") +
#   theme_minimal() +
#   theme(plot.title = element_markdown(size = 14), legend.position = "none")

ggsave("output/ggtext_title.png", p_title, width = 9, height = 5, dpi = 300)

# ── 5. FULL CALLOUT ANNOTATION ───────────────────────────
movies_yr <- yearly_type |> filter(type == "Movie")
peak <- movies_yr |> slice_max(n, n = 1)
latest <- movies_yr |> filter(year_added == 2021)
decline_pct <- round((1 - latest$n / peak$n) * 100)

p_callout <- ggplot(yearly_type, aes(x = year_added, y = n, group = type)) +
  # Grey context
  geom_line(color = "#DDDDDD", linewidth = 0.8) +
  geom_line(data = movies_yr, color = "#1565C0", linewidth = 2.5) +
  geom_point(data = movies_yr, color = "#1565C0", size = 3) +
  # Reference line at peak
  geom_hline(yintercept = peak$n, linetype = "dotted", color = "#888") +
  # Shaded crisis region
  annotate("rect", xmin = 2019.5, xmax = 2021.5, ymin = -Inf, ymax = Inf,
    fill = "#E53935", alpha = 0.03) +
  # Callout box with arrow
  annotate("label",
    x = 2018, y = latest$n,
    label = paste0(decline_pct, "% decline\nfrom 2019 peak"),
    size = 3.5, fontface = "bold", color = "#E53935",
    fill = "white", label.size = 0.5) +
  annotate("segment", x = 2018.8, xend = 2020, y = latest$n, yend = latest$n + 20,
    arrow = arrow(length = unit(0.15, "cm")), color = "#E53935") +
  # Direct labels (no legend)
  geom_text(data = yearly_type |> filter(year_added == 2021),
    aes(label = type, color = type == "Movie"),
    hjust = -0.15, size = 3, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#1565C0", "FALSE" = "#BBBBBB")) +
  coord_cartesian(xlim = c(2015, 2022.5)) +
  theme_minimal(base_size = 11) +
  theme(legend.position = "none",
    plot.title = element_text(face = "bold", size = 13)) +
  labs(title = paste0("Movie additions declined ", decline_pct, "% from their 2019 peak"),
       subtitle = "Shaded region marks the decline period | Dotted line = peak level",
       x = NULL, y = "Titles Added",
       caption = "Source: Netflix dataset | UNYT")
ggsave("output/fully_dressed.png", p_callout, width = 10, height = 6, dpi = 300)

# ── 6. e-Car: ANNOTATED CRISIS CHART ────────────────────
ec <- read_csv("ecar.csv") |>
  mutate(Year = year(mdy(`Approve Date`)),
    Spread = Rate - `Cost of Funds`) |>
  filter(Year >= 2002, Year <= 2012)
yearly_ec <- ec |> group_by(Year) |>
  summarise(rate = mean(Rate), spread = mean(Spread))

p_ecar_ann <- ggplot(yearly_ec) +
  # Shaded crisis window
  annotate("rect", xmin = 2007.5, xmax = 2009.5, ymin = -Inf, ymax = Inf,
    fill = "#E53935", alpha = 0.05) +
  # Lines
  geom_line(aes(x = Year, y = spread), color = "#DDDDDD", linewidth = 0.8) +
  geom_line(aes(x = Year, y = rate), color = "#1565C0", linewidth = 2.5) +
  geom_point(aes(x = Year, y = rate), color = "#1565C0", size = 2.5) +
  # Callout: Lehman
  annotate("label", x = 2008, y = max(yearly_ec$rate) * 0.95,
    label = "Lehman Brothers\ncollapse (Sep 2008)",
    size = 2.5, fontface = "bold", color = "#E53935",
    fill = "white", label.size = 0.5) +
  annotate("segment", x = 2008, xend = 2008,
    y = max(yearly_ec$rate) * 0.88, yend = yearly_ec$rate[yearly_ec$Year == 2008],
    arrow = arrow(length = unit(0.15, "cm")), color = "#E53935") +
  # Direct labels
  annotate("text", x = 2012.3, y = yearly_ec$rate[yearly_ec$Year == 2012],
    label = "Rate", color = "#1565C0", fontface = "bold", size = 3.5, hjust = 0) +
  annotate("text", x = 2012.3, y = yearly_ec$spread[yearly_ec$Year == 2012],
    label = "Spread", color = "#AAAAAA", size = 3, hjust = 0) +
  # Reference lines
  geom_hline(yintercept = 0, linetype = "solid", color = "#888", linewidth = 0.3) +
  scale_x_continuous(breaks = 2002:2012, limits = c(2002, 2013)) +
  theme_minimal(base_size = 10) +
  theme(legend.position = "none",
    plot.title = element_text(face = "bold", size = 12)) +
  labs(title = "Loan rates dropped 2pp after the 2008 financial crisis",
       subtitle = "Shaded region = crisis window | Rate recovered partially by 2012",
       x = NULL, y = "Percentage Points",
       caption = "Source: e-Car dataset | UNYT")
ggsave("output/ecar_annotated.png", p_ecar_ann, width = 10, height = 5, dpi = 300)

# ── 7. ANNOTATION HIERARCHY COMPARISON ───────────────────
# Layer 0: bare chart
p_bare <- ggplot(movies_yr, aes(x = year_added, y = n)) +
  geom_line(color = "#1565C0", linewidth = 1) + geom_point(color = "#1565C0", size = 2) +
  theme_minimal(base_size = 8) +
  labs(title = "Layer 0: Bare chart\n(no annotation)")

# Layer 1: + reference line
p_ref <- p_bare +
  geom_hline(yintercept = peak$n, linetype = "dotted", color = "#888") +
  labs(title = "Layer 1: + Reference line\n(peak level)")

# Layer 2: + direct label
p_label <- p_ref +
  annotate("text", x = 2021.2, y = latest$n, label = "Movies",
    color = "#1565C0", fontface = "bold", size = 3, hjust = 0) +
  labs(title = "Layer 2: + Direct label")

# Layer 3: + callout
p_full <- p_label +
  annotate("label", x = 2018, y = latest$n * 0.85,
    label = paste0("–", decline_pct, "%"), size = 3, fontface = "bold",
    color = "#E53935", fill = "white", label.size = 0.4) +
  labs(title = "Layer 3: + Callout box")

ggsave("output/annotation_layers.png",
  (p_bare | p_ref) / (p_label | p_full) +
  plot_annotation(title = "Annotation Hierarchy: Build in Layers"),
  width = 12, height = 8, dpi = 300)

cat("\n── All W07-M03 R outputs saved ──\n")
